File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part06_canconstruct/java/CountConstructMemoised.java
package part06_canconstruct.java;

import java.util.HashMap;
import java.util.Map;

public class CountConstructMemoised {
  private static Map<String, Integer> countConstructMemmo = new HashMap<>();

  public static void main(String[] args) {
    System.out.println(countConstruct("abcdef", new String[]{"ab", "abc", "cd", "def", "abcd"}));
    countConstructMemmo.clear();

    System.out.println(countConstruct("purple", new String[]{"purp", "p", "ur", "le", "purpl"}));
    countConstructMemmo.clear();

    System.out.println(countConstruct("skateboard", new String[]{"bo", "rd", "ate", "t", "ska", "sk", "boar"}));
    countConstructMemmo.clear();
    System.out.println(countConstruct("enterapotentpot", new String[]{"a", "p", "ent", "enter", "ot", "o", "t"}));
    countConstructMemmo.clear();
    System.out.println(countConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef", new String[]{"e", "ee", "eee", "eeee", "eeeee", "eeeeee"}));
    countConstructMemmo.clear();
  }

  private static int countConstruct(String targetStr, String[] wordBank) {
    if(countConstructMemmo.containsKey(targetStr)){
      return countConstructMemmo.get(targetStr);
    }
    if(targetStr == ""){
      return 1;
    }

    int totalNumWays = 0;

    for(String word : wordBank){
      if(targetStr.indexOf(word) == 0){
        int numWays = countConstruct(targetStr.substring(word.length()), wordBank);
        totalNumWays += numWays;
      }
    }
    countConstructMemmo.put(targetStr, totalNumWays);
    return totalNumWays;
  }
}

File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part06_canconstruct/javascript/countConstructMemoised.js
const countConstruct = (target, wordBank, memo = {}) => {
  if (target in memo) return memo[target];
  if (target === "") return 1;

  let totalCount = 0;

  for (let word of wordBank) {
    if (target.indexOf(word) === 0) {
      const numWaysForRest = countConstruct(target.slice(word.length), wordBank, memo);
      totalCount += numWaysForRest;
    }
  }
  memo[target] = totalCount;
  return totalCount;
};

// Brute force
// time: O(m^n * m)
// space: O(m^2)

// Memoisation
// time: O(n * m^2)
// space: O(m^2)

console.log(countConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"])); // 1
console.log(countConstruct("purple", ["purp", "p", "ur", "le", "purpl"])); // 2
console.log(countConstruct("scateboard", ["bo", "rd", "ate", "t", "ska", "sk", "boar"])); //false
console.log(countConstruct("enterapotentpot", ["a", "p", "ent", "enter", "ot", "o", "t"])); // true
console.log(countConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef", ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"])); //false
